source/common/matcher/field_matcher.h

```diff
@@ -0,0 +1,129 @@
+
+#pragma once
+
+#include "envoy/matcher/matcher.h"
+
+namespace Envoy {
+namespace Matcher {
+
+/**
+ * Base class for matching against a single input.
+ */
+template <class DataType> class FieldMatcher {
+public:
+  virtual ~FieldMatcher() = default;
+
+  /**
+   * Attempts to match against the provided data.
+   * @returns absl::optional<bool> if matching was possible, the result of the match. Otherwise
+   * absl::nullopt if the data is not available.
+   */
+  virtual absl::optional<bool> match(const DataType& data) PURE;
+};
+template <class DataType> using FieldMatcherPtr = std::unique_ptr<FieldMatcher<DataType>>;
+
+/**
+ * A FieldMatcher that attempts to match multiple FieldMatchers, evaluating to true iff all the
+ * FieldMatchers evaluate to true.
+ *
+ * If any of the underlying FieldMatchers are unable to produce a result, absl::nullopt is returned.
+ */
+template <class DataType> class AllFieldMatcher : public FieldMatcher<DataType> {
+public:
+  explicit AllFieldMatcher(std::vector<FieldMatcherPtr<DataType>>&& matchers)
+      : matchers_(std::move(matchers)) {}
+
+  absl::optional<bool> match(const DataType& data) override {
+    for (const auto& matcher : matchers_) {
+      const auto result = matcher->match(data);
+
+      if (!result) {
+        return result;
+      }
+
+      if (!*result) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+private:
+  const std::vector<FieldMatcherPtr<DataType>> matchers_;
+};
+
+/**
+ * A FieldMatcher that attempts to match multiple FieldMatchers, evaluating to true iff any of the
+ * FieldMatchers evaluate to true.
+ *
+ * If any of the underlying FieldMatchers are unable to produce a result before we see a successful
+ * match, absl::nullopt is returned.
+ */
+template <class DataType> class AnyFieldMatcher : public FieldMatcher<DataType> {
+public:
+  explicit AnyFieldMatcher(std::vector<FieldMatcherPtr<DataType>>&& matchers)
+      : matchers_(std::move(matchers)) {}
+
+  absl::optional<bool> match(const DataType& data) override {
+    for (const auto& matcher : matchers_) {
+      const auto result = matcher->match(data);
+
+      if (!result) {
+        return result;
+      }
```

Comment:
Is this correct? Don't you want to keep iterating to see if any of the other matchers are true?